File: backend.py
# pyright: reportAny = false
# pyright: reportExplicitAny = false
# pyright: reportUnusedCallResult = false

import argparse
import functools
import itertools
import json
import logging
import os
from concurrent.futures import ThreadPoolExecutor
from pprint import pprint
from typing import Any, final

import requests

type Media = dict[str, Any]


@final
class APIConfig:
    def __init__(self, timeout_s: int = 15):
        assert timeout_s > 0, "Timeout must be a positive integer"

        API_VERSION = "v24.0"
        INSTAGRAM_ID = "17841455210130919"
        API_URL = f"https://graph.facebook.com/{API_VERSION}/{INSTAGRAM_ID}"

        access_token = os.getenv("ACCESS_TOKEN", "")
        assert access_token, "access token is empty"

        self.url = API_URL
        self.access_token = access_token
        self.timeout_s = timeout_s


def main():
    logging.basicConfig(format="%(asctime)s: [%(levelname)s] %(message)s")

    config = APIConfig()

    parser = init_parser()
    args = parser.parse_args()

    if args.command == "check":
        is_creator_account = check_if_creator_account(args.username, config)
        print(is_creator_account)
    elif args.command == "get":
        if args.all:
            handle_get_all(config)
        else:
            data = get_user_data(args.username, config)
            media = sanitize_data(args.username, data)
            pprint(media)
    else:
        parser.print_help()


def init_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser()
    subparsers = parser.add_subparsers(dest="command")

    # `check` command
    parser_check = subparsers.add_parser(
        "check", help="Verifies if the user is accesible by the API"
    )
    parser_check.add_argument("username", type=str, help="Instagram username")

    # `get` command
    parser_get = subparsers.add_parser("get", help="Gets the media of an user")
    group_get = parser_get.add_mutually_exclusive_group(required=True)
    group_get.add_argument("username", nargs="?", help="Username to get the media from")
    group_get.add_argument(
        "--all",
        action="store_true",
        help="Gets the media from all the users in USERS_PATH",
    )

    return parser


def handle_get_all(config: APIConfig):
    USERS_PATH = "src/lib/assets/users.json"
    MEDIA_PATH = "src/lib/assets/media.json"

    with open(USERS_PATH, "r") as file:
        users: list[str] = json.load(file)["users"]

    media = get_all_users_media(users, config)
    media = sorted(media, key=lambda post: post["timestamp"], reverse=True)
    data = json.dumps(media, indent=2)

    with open(MEDIA_PATH, "w") as file:
        file.write(data)

    print(f"Posts saved on {MEDIA_PATH}")


def get_all_users_media(
    usernames: list[str],
    config: APIConfig,
    amount: int = 5,
) -> list[Media]:
    assert len(usernames) > 0, "List of usernames must not be empty"
    assert amount > 0, "Amount must be a positive integer"

    with ThreadPoolExecutor() as executor:
        task = functools.partial(process_single_user, config=config, amount=amount)
        results = executor.map(task, usernames)

    media: list[Media] = list(itertools.chain.from_iterable(results))
    return media


def process_single_user(
    username: str, config: APIConfig, amount: int = 5
) -> list[Media]:
    user_data = get_user_data(username, config, amount)
    return sanitize_data(username, user_data)


def get_user_data(username: str, config: APIConfig, amount: int = 5) -> dict[str, Any]:
    assert amount > 0, "Amount of media to retrieve must be positive"

    fields = f"""
        business_discovery.username({username}){{
            profile_picture_url,
            media.limit({amount}){{
                timestamp,caption,media_type,permalink,media_url,
                children{{media_url,media_type}}
            }}
        }}
    """
    params = {"fields": fields, "access_token": config.access_token}

    response = None
    try:
        response = requests.get(config.url, params, timeout=config.timeout_s)
        response.raise_for_status()
        return response.json()

    except requests.exceptions.RequestException as error:
        logging.error(f"Could not connect to Instagram API ({error})")
        if response is not None:
            logging.error(response.text)

        return {}


def sanitize_data(username: str, data: dict[str, Any]) -> list[Media]:
    media_list: list[Media] = []

    m: Media
    for m in data["business_discovery"]["media"]["data"]:
        media = m.copy()

        # Shared data between media
        media.pop("id", None)
        media["username"] = username
        media["profile_picture_url"] = data["business_discovery"]["profile_picture_url"]

        if "children" not in m:
            media_list.append(media)
            continue

        # Clean children
        children: list[Media] = m["children"]["data"]
        for child in children:
            child.pop("id", None)
        m["children"] = children

        media_list.append(media)

    return media_list


def check_if_creator_account(username: str, config: APIConfig) -> bool:
    fields = f"business_discovery.username({username}){{name}}"
    params = {"fields": fields, "access_token": config.access_token}

    response = None
    try:
        response = requests.get(config.url, params, timeout=config.timeout_s)
        response.raise_for_status()
        return response.status_code == 200

    except requests.exceptions.RequestException as error:
        logging.error(f"Could not connect to Instragram API ({error})")
        if response is not None:
            logging.error(response.text)

        return False


if __name__ == "__main__":
    main()

File: backend.ts
import { file, write } from "bun";

type User = { username: string; category: string };
type Media = Record<string, any>;

class APIConfig {
  url: string;
  accessToken: string;
  timeoutSeconds: number;

  constructor(timeoutSeconds: number = 15) {
    if (timeoutSeconds <= 0) throw new Error("Timeout must be greater than 0");

    const API_VERSION = "v24.0";
    const INSTAGRAM_ID = "17841455210130919";
    this.url = `https://graph.facebook.com/${API_VERSION}/${INSTAGRAM_ID}`;

    const token = process.env.ACCESS_TOKEN;
    if (!token) throw new Error("ACCESS_TOKEN environment variable is required");

    this.accessToken = token;
    this.timeoutSeconds = timeoutSeconds;
  }
}

main().catch((error) => {
  console.error("Execution failed:", error);
  process.exit(1);
});

async function main() {
  const USERS_PATH = "src/lib/assets/users.json";
  const MEDIA_PATH = "src/lib/assets/media.json";

  const config = new APIConfig();

  const usersFile = file(USERS_PATH);
  const usersData = await usersFile.json();

  const users: User[] = [];
  for (const [catergory, usernames] of Object.entries(usersData)) {
    for (const username of usernames as string[]) {
      users.push({ username, category: catergory });
    }
  }

  let media = await getAllUsersMedia(users, config);

  media.sort((a, b) => {
    const timeA = new Date(a.timestamp).getTime();
    const timeB = new Date(b.timestamp).getTime();
    return timeB - timeA;
  });

  const outputData = {
    lastUpdate: new Date().toISOString(),
    media: media,
  };

  await write(MEDIA_PATH, JSON.stringify(outputData, null, 2));
  console.log(`Posts saved on ${MEDIA_PATH}`);
}

async function getAllUsersMedia(users: User[], config: APIConfig): Promise<Media[]> {
  if (users.length === 0) throw new Error("Users array cannot be empty");

  const promises = users.map((user) => processSingleUser(user, config));
  const results = await Promise.all(promises);
  return results.flat();
}

async function processSingleUser(user: User, config: APIConfig): Promise<Media[]> {
  const userData = await getUserData(user.username, config);
  return sanitizeData(user.username, userData, user.category);
}

async function getUserData(
  username: string,
  config: APIConfig,
  sinceDays: number = 30,
  maxAmount: number = 100,
): Promise<any> {
  if (maxAmount <= 0) throw new Error("maxAmount must be greater than 0");

  const dateLimit = new Date();
  dateLimit.setDate(dateLimit.getDate() - sinceDays);
  const timestamp = Math.floor(dateLimit.getTime() / 1000);

  const fields = `
  business_discovery.username(${username}){
    profile_picture_url,
    media.limit(${maxAmount}).since(${timestamp}){
      timestamp,caption,media_type,permalink,media_url,children
    }
  }`;

  const url = new URL(config.url);
  url.searchParams.append("fields", fields);
  url.searchParams.append("access_token", config.accessToken);

  try {
    const response = await fetch(url.toString(), {
      signal: AbortSignal.timeout(config.timeoutSeconds * 1000),
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error(`Error HTTP ${response.status}: ${errorText}`);
      return {};
    }

    return await response.json();
  } catch (error) {
    console.error(`Could not connecto to Instagram API: ${error}`);
    return {};
  }
}

function sanitizeData(username: string, data: any, category: string = "") {
  if (!data?.business_discovery?.media) {
    return [];
  }

  const profilePictureUrl = data.business_discovery.profile_picture_url;
  return data.business_discovery.media.data.map((m: any) => {
    const media = { ...m };
    delete media.id;
    media.username = username;
    media.category = category;
    media.profile_picture_url = profilePictureUrl;
    media.children = "children" in media;
    return media;
  });
}
